[DataGrid] Allow to customize the value displayed in the filter button tooltip

ithrforu opened the pull request:
Related to https://github.com/mui/mui-x/issues/5492 (a follow-up is to fix the built-in operators)

I faced a problem with current filter values in GridToolbarFilterButton tooltip title (use DataGridPremium) and found [issue](https://github.com/mui/mui-x/issues/5492#issuecomment-1185835877) with @m4theushw comment. This problem looks like:
![image](https://user-images.githubusercontent.com/81531000/203281975-ae88f01f-752b-45b7-8055-772ebc3bd3a2.png)
![image](https://user-images.githubusercontent.com/81531000/203282270-b31355af-55cf-4d6e-a59e-b533db29e030.png)

In this PR i added the ability (from @m4theushw proposal) to specify a function to format the filter value in GridToolbarFilterButton tooltip title.

MBilalShafi commented on packages/grid/x-data-grid/src/components/toolbar/GridToolbarFilterButton.tsx:
```diff
@@ -78,6 +78,14 @@ const GridToolbarFilterButton = React.forwardRef<HTMLButtonElement, GridToolbarF
           .getLocaleText(`filterOperator${capitalize(item.operatorValue!)}` as GridTranslationKeys)!
           .toString();
 
+      const getFilterItemValue = (item: GridFilterItem): string => {
+        const { getValueAsString } = lookup[item.columnField!].filterOperators!.find(
+          (operator) => operator.value === item.operatorValue,
+        )!;
+
+        return getValueAsString ? getValueAsString(item.value) : item.value;
+      };
```
Filter item got [updated](https://github.com/mui/mui-x/pull/6985), can you rebase with latest `next` ?
```suggestion
      const getFilterItemValue = (item: GridFilterItem): string => {
        const { getValueAsString } = lookup[item.field!].filterOperators!.find(
          (operator) => operator.value === item.operator,
        )!;

        return getValueAsString ? getValueAsString(item.value) : item.value;
      };
```

ithrforu replied:
All done.

MBilalShafi commented:
@ithrforu Thanks for the contribution, it seems good, can we test the newly added function `getValueAsString` and add some example usage in demos to improve usability?

ithrforu commented:
@MBilalShafi. Thank you for attention. Of course. Should I create new pull request or continue here?

MBilalShafi commented:
> Should I create new pull request or continue here

It will be better if we have it as part of this pull request so that we push this new addition tested and more usable.

ithrforu commented:
@MBilalShafi, I added an example and rebased with latest next, but got error in [docs](https://deploy-preview-6956--material-ui-x.netlify.app/x/react-data-grid/filtering/) page.
![image](https://user-images.githubusercontent.com/81531000/210170062-1ded120d-79bf-4f63-bb64-b7581aa2368f.png)

How can i test `getValueAsString`?

MBilalShafi commented:
> How can I test `getValueAsString`?

You can create a new test case in https://github.com/mui/mui-x/blob/next/packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx covering this use-case and test if it works as expected.

ithrforu commented:
@MBilalShafi, all done.

MBilalShafi commented on docs/data/data-grid/filtering/filtering.md:
```diff
@@ -203,6 +203,10 @@ The [`valueFormatter`](/x/react-data-grid/column-definition/#value-formatter) is
 :::
 
+:::info
+The `getValueAsString` function allows you to convert the value of a filter item to a human-readable form before displayed in the UI of filter button tooltip.
```
```suggestion
The filter button displays a tooltip on hover if there are active filters. Pass [`getValueAsString`](/x/api/data-grid/grid-filter-operator/) in the filter operator to customize or convert the value to a more human-readable form.
```

ithrforu replied:
Thanks for your edits and remarks. Text is fixed.

ithrforu commented:
Is it bug?
![MUI](https://user-images.githubusercontent.com/81531000/216048640-9c4ee3fc-0b43-4577-a8a9-65b8baca5a1b.gif)

ithrforu commented:
@m4theushw, I mean I hovered filter button but "Show the code" tooltip displayed.

LukasTy commented:
Indeed, this PR might have some sort of regression in hover event bubbling or something along these lines. The current versions do not display `Show the full source` tooltip on demo hover. 🙈

LukasTy commented:
> I can also reproduce this bug in https://material-ui.netlify.app/material-ui/react-button/ so certainly is not a regression introduced by this PR but from the core.

Hm. Good point. I tried testing a local dev run and did not see this behaviour, that's why I thought that it might have something to do with this PR, but I do not see any `@mui/monorepo` bumps to warrant that. 🙈

ithrforu commented:
Can i do something?

MBilalShafi commented:
It's all good @ithrforu, I'll merge this now.
Regarding the above-mentioned regression, I'll create a ticket in core if there's not one already.

LukasTy commented:
> It's all good @ithrforu, I'll merge this now. Regarding the above-mentioned regression, I'll create a ticket in core if there's not one already.

It might have also been a desired change from MUI Core, we just weren't aware of it. But I couldn't find a recently merged PR related to it. 🤔
